### ISSUE TITLE
Terraform sparse checkout module

### ISSUE THREAD

**Issue by rverma-nikiai** [hashicorp/terraform#19277]
### Current Terraform Version
```
v0.11.7
```

### Use-cases
Terraform module sparse checkout and specify depth. While terraform suggest 1 module per repo, there are orgs which are more willing to manage multiple related modules together. This gives faster feedback cycle also, related pull requests in one repo etc..

### Attempted Solutions
Couldn't find any thing relevant.

### Proposal
possibly we can evolve source with backward compatibility as 

```
module dynamo-auto {
     source = "git::https://github.com/cloudposse/terraform-aws-dynamodb-autoscaler.git?ref=master"}
```

and also 

```
module dynamo-auto {
     source = {
       repo = "git::https://github.com/cloudposse/terraform-aws-dynamodb-autoscaler.git?ref=master"
       depth = 1
       path = /modules/dynamo
}
```
which allows to sparse checkout /modules/dynamo as relevant terraform module.

**Comment by apparentlymart** [comment]
Hi @rverma-nikiai! Thanks for sharing this use-case.

The git module source actually already supports a syntax for selecting a sub-path from a repository, like this:

```hcl
module "dynamo-auto" {
     source = "git::https://github.com/cloudposse/terraform-aws-dynamodb-autoscaler.git//modules/dynamo?ref=master"
}
```

The extra `//...` portion of the path at the end is interpreted as a subdirectory within the repository.

That then just leaves the request for shallow cloning. The git handling is all done by a component which parses only the `source` string, so additional git-related settings must be packed in inside that pseudo-query-string argument at the end, which means a hypothetical new option might look something like this:

```hcl
module "dynamo-auto" {
     source = "git::https://github.com/cloudposse/terraform-aws-dynamodb-autoscaler.git//modules/dynamo?ref=master&depth=1"
}
```

However, I think Terraform's module installer does a full clone by default just because when it was written the "shallow clone" functionality was relatively new and limited, and we wanted to be sure of proper behavior on subsequent commands such as upgrading the module, which requires running operations like `git fetch`.

I think we should investigate whether the improved shallow clone behavior added in Git 1.9 (now several years old) is featureful enough that we could enable shallow cloning _by default_ in a future release, since the module installer's goal is always to install just the single version you requested, rather than to create a fully-fledged development environment for that repository. Before making that decision, we'll need to prototype it to make sure the upgrading behavior is well-behaved after a shallow, single-branch clone.

We are in the early stages of planning some other changes to how Terraform manages configuration dependencies for a future release, so I'm going to label this one to remind us to consider this use-case as part of that work.

**Comment by rifelpet** [comment]
@apparentlymart Now that https://github.com/hashicorp/go-getter/pull/140 has been merged, any chance we can get terraform's vendoring updated to add support for shallow clones? I'm happy to open a PR including a docs update, I just need to know which target branch would be most appropriate at the moment.

**Comment by rifelpet** [comment]
It looks like #20411 updated the go-getter version to include the shallow clone functionality. It is in 0.12 beta1. I'm looking forward to using this in 0.12, thanks! We can probably close this issue out once 0.12 is released.

**Comment by apparentlymart** [comment]
Since that other PR wasn't intentionally updating `go-getter` to address this issue, it therefore didn't update Terraform's module sources documentation to mention this new option. We'll need to do _that_ at least before considering this done.

I'd also still like to investigate whether that option is necessary at all or if we can just make that behavior the default. Since we're not cloning the repository for _development_ it seems unnecessary to produce a fully-functioning work tree by default, and in the rare case where someone _does_ want to work directly with the cloned repository in `.terraform/modules` it only takes a couple git commands to fetch the full history if needed.

**Comment by rifelpet** [comment]
Ah you're right I forgot about documentation, and I would fully support using depth=1 by default. I can't think of any reasonable situations where a user would need the full history in `.terraform/modules`.

**Comment by KaGeN101** [comment]
Surely the source can be cached as well as put in a lookup during execution. If the lookup contains the same url key just use the cached copy if not source it and add to the cache...this surely can't be hard to do versus downloading and identical copy form the internet over and over and over each time :/

**Comment by adrian-gierakowski** [comment]
@apparentlymart 

Is there a github issue tracking the the following point?

> 2\. Terraform clones the entire history of the repository, even though it only uses the latest commit on the given branch.

I'm using the [terraform-google-modules/gcloud/google](https://github.com/terraform-google-modules/terraform-google-gcloud) module which ends up downloading hundreds of megabytes of history since the github repo contains `gcloud` binaries and grows in size significantly with every version bump.

**Comment by KaGeN101** [comment]
> @apparentlymart
> 
> Is there a github issue tracking the the following point?
> 
> > 2. Terraform clones the entire history of the repository, even though it only uses the latest commit on the given branch.
> 
> I'm using the [terraform-google-modules/gcloud/google](https://github.com/terraform-google-modules/terraform-google-gcloud) module which ends up downloading hundreds of megabytes of history since the github repo contains `gcloud` binaries and grows in size significantly with every version bump.

That is the way Git works it is distributed so copies all history locally. There no way around this the first time

**Comment by milpog** [comment]
It would be really nice to have shallow clone as default option for cloning Terraform modules.

**Comment by apparentlymart** [comment]
Nobody on the Terraform team at HashiCorp is currently working on this, because our attentions are currently elsewhere.

As I mentioned before, the main trick here is making sure that shallow clone won't break the ability for `terraform init -upgrade` to roll forward to a newer commit when a shallow tree is already present on disk. I don't know yet how that will behave, and I think understanding that behavior is the main blocker for deciding whether we can make this change. If someone is motivated to work on this, I'd suggest the following approach to get into a state where it's possible to test and experiment:

* Create a local branch of [`go-getter`](https://github.com/hashicorp/go-getter), which is the library that implements the Git fetching in Terraform.
* In your Terraform work tree, temporarily edit `go.mod` to include a `replace` directive referring to your local `go-getter` tree, so your local Terraform builds will see the `go-getter` changes you're making locally:

    ```gomod
    replace github.com/hashicorp/go-getter => ../go-getter
    ```
* Change the logic in [`GitGetter`](https://github.com/hashicorp/go-getter/blob/master/get_git.go) to enable shallow cloning unconditionally. (I don't have exact details on this step, because I've not looked closely at the logic in there yet.)

* Build Terraform against the locally-modified `go-getter` and experiment with `terraform init` and `terraform init -upgrade` to make sure they are both still working as expected.

If the above is fruitful and it seems like making shallow clone the default work work, I expect the final change to `go-getter` would need to make it conditional via a flag field in [the `GitGetter` type](https://github.com/hashicorp/go-getter/blob/master/get_git.go#L25) so that Terraform can enable it without forcing that behavior on other `go-getter` callers. We can then change [Terraform's own instantiation of that getter](https://github.com/hashicorp/terraform/blob/6cf3f766231755e12b4fffb1049af775db26da93/configs/configload/getter.go#L50) to set the new flag, making that behavior always be activated for Terraform's module installer.

If someone is interested in working on this but needs some more guidance, please let me know what specific questions you have and I can try to answer them as best I can with what I know already.

**Comment by hayorov** [comment]
Due to closed https://github.com/hashicorp/terraform/issues/11435 I'd like to slightly offtopic here and share a small pre-terraform routine utility that optimizes init (modules download) for git modules https://github.com/hayorov/terraform-init-booster

**Comment by sebglon** [comment]
Can we have some news?

**Comment by sebglon** [comment]
It seems that we have a beginning fix here: https://github.com/hashicorp/terraform/issues/10703

**Comment by Vasi-Shche** [comment]
> 3. Terraform clones the entire source tree in the repository, even though only a sub-path is requested.

> Point 3 here is intentional because in a multi-module repository the different modules will often refer to one another with references like `source = "../es"` and so we need to have the entire repository on disk to resolve references like that.

About Point 3. With 100 modules per repo, it will replicate it 100 times. Why not, for example, use links instead of copying 100 modules 100 times?

**Comment by aberres** [comment]
I wanted to reference modules living in a mono repo. But boy, this is no fun.

**Comment by aberres** [comment]
@apparentlymart But then we use the possibility to version things, don't we? Or is there a trick?

**Comment by apparentlymart** [comment]
Indeed, when I hear "monorepo" I tend to understand that to mean "big bag of everything all versioned together as a single unit", so I was assuming that independent module versioning is not a requirement. If you _do_ want to have separate versions for each of your modules then indeed it'd be best to use multiple smaller repositories to represent that, so that the boundaries between the packages are clear.

**Comment by aberres** [comment]
> Indeed, when I hear "monorepo" I tend to understand that to mean "big bag of everything all versioned together as a single unit"

You are absolutely right. Nevertheless, I was hoping to leverage versioning to do things like promoting things from testing, to staging, to prod.
Creating (TF specific) tags works even in the monorepo case.

> If you do want to have separate versions for each of your modules then indeed it'd be best to use multiple smaller repositories to represent that, so that the boundaries between the packages are clear.

Yeah, I guess we should simply move the TF code out and call it a day 👍

**Comment by fishpen0** [comment]
You can use git tags to pull off versioning inside a monorepo. So for each module in our modules folder we autogenerate a semver tag that looks like this: 'foomod-1.1.1' from a 'version' file we keep collocated in the module folder. Then if we want to use that module we reference it like so from our same repo:

'''
module "myfoomod" {
  src = "git@github.com:myorg/monorepo.git//modules/foomod?ref=foomod-1.1.1"
} 
'''

It is not ideal though because if you do this a lot in a monorepo you get hundreds of local copies of the modules cloning the repo you already are working inside of over and over again. Which is kind of what this issue is opened to fix.

**Comment by chernetskyi** [comment]
> As I mentioned before, the main trick here is making sure that shallow clone won't break the ability for `terraform init -upgrade` to roll forward to a newer commit when a shallow tree is already present on disk. I don't know yet how that will behave, and I think understanding that behavior is the main blocker for deciding whether we can make this change. If someone is motivated to work on this, I'd suggest the following approach to get into a state where it's possible to test and experiment:

@apparentlymart do we really need such a complicated experiment to test unconditional shallow cloning?

I have just ran `terraform init` with shallow cloned module with pinned branch, updated the branch, and ran `terraform init -upgrade` to verify if the module will be updated. It will. It should come as no surprise, since shallow cloning only truncates the history, and does not affect the fetching of new commits in any way.

Do you or anyone else have any other concerns, why shallow cloning should not be unconditional?

**Comment by apparentlymart** [comment]
Hi @chernetskyi,

I don't work on Terraform anymore, but hopefully your feedback will be helpful to the current team. Thanks!

**Comment by kumadee** [comment]
Refer https://developer.hashicorp.com/terraform/language/block/module#specify-depth-of-clone

Just like ref, depth is another http request parameter.
```hcl
module "gke_cluster" {
  source = "git::https://github.com/GoogleCloudPlatform/cloud-foundation-fabric.git//modules/gke-cluster-standard?depth=1&ref=v41.0.0"
}
```
